webapp/src/main/kotlin/serialization/MessageSerializer.kt

```diff
@@ -0,0 +1,147 @@
+        override val code: Code,
+        val id: Int,
+        override val token: Long,
+        override val options: List<Option>,
+    ) : MessageSurrogate()
+
+    @Serializable
+    data class Tcp(
+        override val code: Code,
+        override val token: Long,
+        override val options: List<Option>,
+    ) : MessageSurrogate()
+}
+
+internal object UdpMessageSerializer : KSerializer<Message.Udp> {
+
+    override val descriptor: SerialDescriptor = MessageSurrogate.Udp.serializer().descriptor
+
+    override fun serialize(
+        encoder: Encoder,
+        value: Message.Udp
+    ) {
+        val surrogate = MessageSurrogate.Udp(
+            type = value.type,
+            code = value.code,
+            id = value.id,
+            token = value.token,
+            options = value.options,
+        )
+
+        encoder.encodeSerializableValue(
+            serializer = MessageSurrogate.Udp.serializer(),
+            value = surrogate
+        )
+    }
+
+    override fun deserialize(decoder: Decoder): Message.Udp = TODO("Not yet implemented")
+}
+
+internal object TcpMessageSerializer : KSerializer<Message.Tcp> {
+
+    override val descriptor: SerialDescriptor = MessageSurrogate.Tcp.serializer().descriptor
+
+    override fun serialize(
+        encoder: Encoder,
+        value: Message.Tcp
+    ) {
+        val surrogate = MessageSurrogate.Tcp(
+            code = value.code,
+            token = value.token,
+            options = value.options,
+        )
+
+        encoder.encodeSerializableValue(
+            serializer = MessageSurrogate.Tcp.serializer(),
+            value = surrogate
+        )
+    }
+
+    override fun deserialize(decoder: Decoder): Message.Tcp = TODO("Not yet implemented")
+}
+
+private object TypeSerializer : KSerializer<Type> {
+
+    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("Type", PrimitiveKind.STRING)
+
+    override fun serialize(encoder: Encoder, value: Type) {
+        val serialized = when (value) {
+            Confirmable -> "Confirmable"
+            NonConfirmable -> "NonConfirmable"
+            Acknowledgement -> "Acknowledgement"
+            Reset -> "Reset"
+        }
+        encoder.encodeString(serialized)
+    }
+
+    override fun deserialize(decoder: Decoder): Type = TODO("Not yet implemented")
+}
+
+private object CodeSerializer : KSerializer<Code> {
+
+    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("Code", PrimitiveKind.STRING)
+
+    override fun serialize(encoder: Encoder, value: Code) {
+        encoder.encodeString(value.toString())
+    }
+
+    override fun deserialize(decoder: Decoder): Code = TODO("Not yet implemented")
+}
+
+private object OptionSerializer : KSerializer<Option> {
+
+    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("Option", PrimitiveKind.STRING)
+
+    override fun serialize(encoder: Encoder, value: Option) {
+        val serialized = when (value) {
```

Comment:
seems like the when here and below could be simplified into one?

Reply:
Tried a few things but couldn't figure out how to simplify it further?

Reply:
might just be personal preference and thus I would never enforce it on ya but I think

`when (value) {
	Option.ContentFormat.PlainText -> "Content-Format:text/plain; charset=utf-8"
	...`

would be cleaner and easier to read?

Reply:
Oh, gotchya. Ya, I agree with that.

Unfortunately I snuck some other changes in to properly handle detecting encoding by also reading the `Accept` option. It muddies up the entire thing. So it has gotten slightly messy, but just being a debug tool I think that's acceptable.

**tl;dr** code has changed enough that original comment isn't really applicable anymore.